### PR TITLE
ESQL: shallow copy in EXPAND via shared vecs

### LINKED ISSUE #100528: Block and Vector creation can start to use a block factory that performs circuit breaking - 8.12
Extends https://github.com/elastic/elasticsearch/issues/99173

### Description

This meta issue lists a number of places in the ESQL runtime where Block and Vector creation can start to use a block factory that performs breaking/accountancy that needs to land in 8.12.

- [ ] #100678
- [x] #100640
- [x] #102683
 - [x] Reference tracking https://github.com/elastic/elasticsearch/pull/100408
 - [ ] Remove copy operations
   - [x] Deep copy on aliases https://github.com/elastic/elasticsearch/pull/100408
   - [x] Deep copy in eval https://github.com/elastic/elasticsearch/pull/100408
   - [ ] ~Copy in filter @alex-spies~ **Update:** not planned, currently not clear whether this really improves performance.
   - [x] Copy in expand https://github.com/elastic/elasticsearch/pull/103681
 - [x] operators (all not covered in #99826
   - [x] Enrich @dnhatn 
   - [x] LuceneSource @nik9000 #101396
   - [x] doc values reading @nik9000  #101383
   - [x] stored field reading @nik9000  #101383
   - [x] dissect #101400
 - [x] generate `toString` for `ExpressionEvaluator.Factory` and remove `ThrowingDriverContext` from `EvalOperatorFactory.describe` #100516
 - [x] Move BlockFactory.getNonBreakingInstance() to tests only (https://github.com/elastic/elasticsearch/pull/103661)
 - [ ] Make sure that many many many small pages don't fill up memory without filling the breaker. See `testTooManyEval`.
- [x] Remove Block.Ref (https://github.com/elastic/elasticsearch/pull/102513)
- [x] Remove deprecated Block methods
- [x] BytesRef does allocation that can’t be tracked when going across a page boundary (#102350)

### PR DESCRIPTION
**Pull request by alex-spies**
Use shared vectors to avoid allocations when expanding.

Partially closes https://github.com/elastic/elasticsearch/issues/100528

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/BooleanBigArrayBlock.java [hunk 1]
```diff
@@ -80,21 +100,21 @@ public BooleanBlock expand() {
...
-            return builder.mvOrdering(MvOrdering.DEDUPLICATED_AND_SORTED_ASCENDING).build();
+        vector.incRef();
+        if (nullsMask == null) {
+            return vector.asBlock();
         }
+        BooleanBigArrayBlock expanded = new BooleanBigArrayBlock(
+            vector,
+            vector.getPositionCount(),
+            null,
+            // TODO: we probably need to adjust the breaker before computing the shifted null mask
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/BooleanBigArrayBlock.java:111]: This should be addressed before merging.

Theoretically also applies to the `X-ArrayBlock`s, but the null masks of those are likely much smaller.

Comment (alex-spies) [reply x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/BooleanBigArrayBlock.java:110]: We could probably just pre-adjust the breaker using the existing null mask to estimate the size of the expanded one.

Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/BooleanBigArrayBlock.java:110]: I think the nulls mask will either be `null` itself or will be less than or equal to `the length of the vector / 8`. So it's not *too* bad to pre-adjust and correct the adjustment with the real size.

Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/BooleanBigArrayBlock.java:110]: No need to be precise on the pre-adjust. I think. Better to over-allocate and then adjust back down.

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/BooleanArrayBlock.java [hunk 2]
```diff
@@ -79,21 +98,20 @@ public BooleanBlock expand() {
...
         }
+        BooleanArrayBlock expanded = new BooleanArrayBlock(
+            vector,
+            vector.getPositionCount(),
+            null,
+            shiftNullsToExpandedPositions(),
+            MvOrdering.DEDUPLICATED_AND_SORTED_ASCENDING,
+            blockFactory()
+        );
+        blockFactory().adjustBreaker(expanded.ramBytesUsedOnlyBlock(), true);
```
Comment (dnhatn) [inline x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/BooleanArrayBlock.java:113]: I think we should decRef the vector if we hit the circuit breaker here.

**Review by dnhatn — APPROVED** [review]
I left a comment, but this looks great. Thanks Alex!

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/BooleanArrayBlock.java (hunk 2 above)
Comment (alex-spies) [reply x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/BooleanArrayBlock.java:113]: Oh, absolutely. Good catch, thanks!

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/BooleanBigArrayBlock.java (hunk 1 above)
Comment (alex-spies) [reply x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/BooleanBigArrayBlock.java]: Alright, I'm using the estimation from BlockRamUsageEstimator, which is essentially `length/8`. That should do the trick.

**Review by dnhatn — APPROVED** [review]
LGTM. Thanks for an extra iteration @alex-spies.

**Comment by alex-spies** [comment]
Thanks for the reviews and remarks @dnhatn and @nik9000 !

**Pull request merged**